src/parquet/column/scanner-test.cc

```diff
@@ -0,0 +1,271 @@
+#include "parquet/schema/types.h"
+#include "parquet/util/test-common.h"
+
+using std::string;
+using std::vector;
+using std::shared_ptr;
+
+namespace parquet_cpp {
+
+using schema::NodePtr;
+
+bool operator==(const Int96& a, const Int96& b) {
+  return a.value[0] == b.value[0] &&
+    a.value[1] == b.value[1] &&
+    a.value[2] == b.value[2];
+}
+
+bool operator==(const ByteArray& a, const ByteArray& b) {
+  return a.len == b.len && 0 == memcmp(a.ptr, b.ptr, a.len);
+}
+
+static int FLBA_LENGTH = 12;
+bool operator==(const FixedLenByteArray& a, const FixedLenByteArray& b) {
+  return 0 == memcmp(a.ptr, b.ptr, FLBA_LENGTH);
+}
+
+namespace test {
+
+template <int TYPE, typename T>
+class TestFlatScanner : public ::testing::Test {
+ public:
+  void InitValues() {
+    random_numbers(num_values_, 0, std::numeric_limits<T>::min(),
+        std::numeric_limits<T>::max(), values_.data());
+  }
+
+  void MakePages(const ColumnDescriptor *d, int num_pages, int levels_per_page) {
+    num_levels_ = levels_per_page * num_pages;
+    num_values_ = 0;
+    uint32_t seed = 0;
+    int16_t zero = 0;
+    int16_t max_def_level = d->max_definition_level();
+    int16_t max_rep_level = d->max_repetition_level();
+    vector<int> values_per_page(num_pages, levels_per_page);
+    // Create definition levels
+    if (max_def_level > 0) {
+      def_levels_.resize(num_levels_);
+      random_numbers(num_levels_, seed, zero, max_def_level, def_levels_.data());
+      for (int p = 0; p < num_pages; p++) {
+        int num_values_per_page = 0;
+        for (int i = 0; i < levels_per_page; i++) {
+          if (def_levels_[i + p * levels_per_page] == max_def_level) {
+            num_values_per_page++;
+            num_values_++;
+          }
+        }
+        values_per_page[p] = num_values_per_page;
+      }
+    } else {
+      num_values_ = num_levels_;
+    }
+    // Create repitition levels
+    if (max_rep_level > 0) {
+      rep_levels_.resize(num_levels_);
+      random_numbers(num_levels_, seed, zero, max_rep_level, rep_levels_.data());
+    }
+    // Create values
+    values_.resize(num_values_);
+    InitValues();
+    //values_.insert(values_.begin(), draws_, draws_ + num_values_);
+    Paginate<TYPE, T>(d, values_, def_levels_, max_def_level,
+        rep_levels_, max_rep_level, levels_per_page, values_per_page, pages_);
+  }
+
+  void InitScanner(const ColumnDescriptor *d) {
+    std::unique_ptr<PageReader> pager(new test::MockPageReader(pages_));
+    scanner_ = Scanner::Make(ColumnReader::Make(d, std::move(pager)));
+  }
+
+  void CheckResults(int batch_size) {
+    TypedScanner<TYPE>* scanner = reinterpret_cast<TypedScanner<TYPE>* >(scanner_.get());
+    T val;
+    bool is_null;
+    size_t j = 0;
+    scanner->SetBatchSize(batch_size);
+    for (size_t i = 0; i < num_levels_; i++) {
+      ASSERT_TRUE(scanner->NextValue(&val, &is_null)) << i <<"NV"<< j;
+      if (!is_null) {
+        ASSERT_EQ(values_[j++], val) << i <<"V"<< j;
```

Comment:
We still aren't verifying that the repetition/definition levels scanned are correct per the discussion in #50 . 

I think the right way to handle this may be to instead change the scanner API to have a method:

```c++
bool Next(T* val, int16_t* rep_level, int16_t* def_level, bool* is_null)
```

what do you think?